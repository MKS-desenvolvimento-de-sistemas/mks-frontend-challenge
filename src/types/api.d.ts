export interface IProduct {
  id: number;
  name: string;
  brand: string;
  description: string;
  price: string;
  photo: string;
  createdAt: string;
  updatedAt: string;
}

export interface IProductCart extends IProduct {
  quantity: number;
}

export interface IProducts {
  products: IProduct[];
}
