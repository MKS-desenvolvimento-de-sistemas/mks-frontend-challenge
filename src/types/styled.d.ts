import {} from "styled-components";
import { theme } from "../styles/themes/";

declare module "styled-components" {
  export interface DefaultTheme extends theme {}
}
