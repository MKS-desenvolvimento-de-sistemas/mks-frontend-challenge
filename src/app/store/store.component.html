<div class="store">
    <h1>Store</h1>
    <div *ngIf="isLoading">Loading...</div>
    <app-product-card *ngFor="let product of products" [product]="product" (addToCartEvent)="addToCart($event)"></app-product-card>
  </div>
  