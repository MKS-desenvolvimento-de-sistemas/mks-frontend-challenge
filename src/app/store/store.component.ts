import { Component, OnInit } from '@angular/core';
import { ApiService, Product } from '..//services/api.service';

@Component({
  selector: 'app-store',
  templateUrl: './store.component.html',
  styleUrls: ['./store.component.scss']
})
export class StoreComponent implements OnInit {
  products: Product[] = [];
  isLoading = true;

  constructor(private apiService: ApiService) {}

  ngOnInit() {
    this.apiService.fetchProducts().then(products => {
      this.products = products;
      this.isLoading = false;
    });
  }

  addToCart(product: Product) {
    // Implementar a l√≥gica de adicionar ao carrinho
    console.log('Adicionar ao carrinho:', product);
  }
}
