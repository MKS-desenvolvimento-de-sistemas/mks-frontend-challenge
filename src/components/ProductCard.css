.product-card {
    border: 1px solid #e0e0e0;
    border-radius: 8px;
    overflow: hidden;
    width: 240px;
    text-align: center;
    height: 100%;
    box-shadow: 0 2px 5px rgba(0,0,0,0.1);
    transition: transform 0.2s;
  }
  
  .product-card:hover {
    transform: scale(1.05);
  }
  
  .product-image {
    width: 180px;
    height: 180px;
    object-fit: cover;
  }
  
  .product-info {
    padding: 16px;

  }
  
  .product-name {
    font-size: 1.2em;
    margin: 0;
    color: #333;
    text-align:start ;
  }
  
  .product-description {
    font-size: 0.9em;
    color: #777;
    margin: 8px 0;
  }

  
  .product-price {
    font-size: 1.2em;
    font-weight: bold;
    color: #fff;   
    text-align: end;
  }
  
  .price{
    padding: 5px;
    background-color: #333;
    border-radius: 5px;

  }
  .product-button {
    background-color: #0F52BA;
    color: white;
    border: none;
    padding: 8px 16px;
    border-radius: 0px, 0px, 8px, 8px;
    cursor: pointer;
    transition: background-color 0.2s;
    width: 100%;
  }
  .product-button img{
    width: 12px;
    margin-right: 14px;
  }
  .product-button:hover {
    background-color: #0056b3;
  }
  .name-price{
    display: flex;
    align-items: center;
    justify-content: space-between;
  }