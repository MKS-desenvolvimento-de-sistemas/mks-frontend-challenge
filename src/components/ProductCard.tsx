import { connect } from "react-redux";
import React from "react";
import { addProduct, updateTotalValue } from "../redux/actions";
import shopping from "../images/shopping-bag.svg";
interface Product {
  img: string;
  title: string;
  description: string;
  price: string;
  id: number;
}

interface Props extends Product {
  addProduct: Function;
  updateValue: Function;
}

interface Props {}

class ProductCard extends React.Component<Props> {
  addProductToCart = ({ target }: { target: any }): void => {
    const { id, price } = this.props;
    const { addProduct, updateValue } = this.props;
    addProduct(`${id}`);
    updateValue(Number(price.split(".")[0]));
  };
  render() {
    const { img, title, description, price, id } = this.props as Product;
    return (
      <div className="product-card">
        <img src={img} alt={title} />
        <div className="price-name">
          <p>{title}</p>
          <p data-testid={`product-price-${id}`} className="price">
            R${price.split(".")[0]}
          </p>
        </div>
        <p className="description">{description}</p>
        <button
          data-testid={`buy-${id}`}
          className="comprar"
          id={`${id}`}
          onClick={this.addProductToCart}
          type="button"
        >
          <img src={shopping} />
          Comprar
        </button>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch: any) => ({
  addProduct: (id: number): void => dispatch(addProduct(id)),
  updateValue: (value: number) => dispatch(updateTotalValue(value)),
});

export default connect(null, mapDispatchToProps)(ProductCard);
