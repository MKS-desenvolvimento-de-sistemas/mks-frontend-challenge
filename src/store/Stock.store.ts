import {createSlice} from '@reduxjs/toolkit';

const carrinho = createSlice({
	name: 'carrinho',
	initialState: {
		counter: 0,
		cartOpen: 'none',
	},
	reducers: {
		increment(state) {
			state.counter += 1;
		},
		decrement(state) {
			state.counter -= 1;
		},
		abrirCarrinho(state) {
			state.cartOpen = 'flex';
		},
		fecharCarrinho(state) {
			state.cartOpen = 'none';
		},
	},
});

export const {
	decrement,
	increment,
	abrirCarrinho,
	fecharCarrinho,
} = carrinho.actions;
export default carrinho.reducer;
