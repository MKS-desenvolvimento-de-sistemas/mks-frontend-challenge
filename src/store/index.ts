import { createSlice } from "@reduxjs/toolkit";
import { configureStore } from "@reduxjs/toolkit";
import { ICartItem } from "../pages/Cart";
const initialState = { items: [] } as any;

const cartSlice = createSlice({
  name: "carrinho",
  initialState,
  reducers: {
    addToCart(state, action) {
      state.items.push(action.payload);
    },

    addAdditional(state, action) {
      if (!state.items.some((item) => item.id === action.payload.id)) return;
      const item = state.items.find((item) => item.id === action.payload.id);

      state.items.push(item);
    },
    decreaseItem(state, action) {
      const remainingItems = state.items.filter(
        (item) => item.id === action.payload.id
      );

      if (remainingItems.length === 1) return;

      const itemIndex = state.items.findIndex(
        (el) => el.id === action.payload.id
      );

      state.items.splice(itemIndex, 1);
    },
    deleteItem(state, action) {
      const arr = state.items.filter((item) => item.id !== action.payload.id);
      state.items = arr;
      // do {
      //   const itemIndex = state.items.findIndex(
      //     (el) => el.id === action.payload.id
      //   );

      //   state.items.splice(itemIndex, 1);
      // } while (state.items.some((item) => item.id === action.payload.id));
    },
    deleteAllItems(state) {
      state.items = [];
    },
  },
});

const store = configureStore({
  reducer: cartSlice.reducer,
});

export const cartActions = cartSlice.actions;

export default store;
