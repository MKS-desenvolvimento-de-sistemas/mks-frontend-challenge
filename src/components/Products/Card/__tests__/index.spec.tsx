import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import ProductCard from "../";
import renderer from "react-test-renderer";
import { IProduct } from "../../../../types/api";
import { Provider } from "react-redux";
import store from "../../../../redux/";

const productExample: IProduct = {
  id: 1,
  name: "EXEMPLO 1",
  brand: "Exemplo 1",
  description: "Exemplo 1",
  price: "200",
  photo: "/vercel.svg",
  updatedAt: "2022",
  createdAt: "2022",
};

const formatterDescription =
  productExample.description.length > 63
    ? productExample.description.slice(0, 64) + "..."
    : productExample.description;

describe("ProductCard Component", () => {
  it("should render the product card correctly", () => {
    render(
      <Provider store={store}>
        <ProductCard {...productExample} />
      </Provider>
    );

    const card = screen.getByTestId("card");
    const cardImage = screen.getByTestId("card--image");

    expect(cardImage).toHaveStyle({
      backgroundImage: `url(${productExample.photo})`,
    });

    expect(card).toHaveTextContent(productExample.name);
    expect(card).toHaveTextContent(formatterDescription);
    expect(card).toHaveTextContent(`R$${productExample.price}`);
    expect(card).toHaveTextContent("Comprar");
  });

  it("should match the snapshot", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <ProductCard {...productExample} />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
