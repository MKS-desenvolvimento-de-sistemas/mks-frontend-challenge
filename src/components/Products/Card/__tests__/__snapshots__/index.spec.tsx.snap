// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard Component should match the snapshot 1`] = `
<li
  className="sc-bczRLJ cyfZqA"
  data-testid="card"
>
  <div
    className="sc-gsnTZi hlgEBv"
  >
    <div
      className="sc-dkzDqf gDJlNm"
      data-testid="card--image"
    />
    <div
      className="sc-hKMtZM jwyXmD"
    >
      <h4>
        EXEMPLO 1
      </h4>
      <div
        className="sc-ftvSup LlaOC"
        data-testid="price"
        type="product"
      >
        R$200
      </div>
      <p>
        Exemplo 1
      </p>
    </div>
  </div>
  <button
    className="sc-idiyUo eimsFO"
    data-testid="buy-button"
    onClick={[Function]}
    title="Adicionar ao Carrinho"
    type="button"
  >
    <svg
      fill="none"
      height="16px"
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      style={
        {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="16px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 2L3 6v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V6l-3-4z"
      />
      <line
        x1="3"
        x2="21"
        y1="6"
        y2="6"
      />
      <path
        d="M16 10a4 4 0 0 1-8 0"
      />
    </svg>
    Comprar
  </button>
</li>
`;
